orttraining/orttraining/python/training/ortmodule/hierarchal_ortmodule.py

```diff
@@ -0,0 +1,163 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+# debug_options.py
+import torch
+from . import ORTModule
+
+
+class HierarchalORTModule(torch.nn.Module):
+    def __init__(self, module):
+        super(HierarchalORTModule, self).__init__()
+        self._initialized = False
+        self._module = module
+
+    def _initialize(self, *args, **kwargs):
+        handle_pool = []
+        module_arg_pool = {}
+
+        # A forward pre-hook to record inputs for each nn.Module.
+        def record_args(module, args):
+            if module in module_arg_pool:
+                # Maybe throw if one module is called multiple times.
+                module_arg_pool[module].append(args)
+            else:
+                module_arg_pool[module] = [args]
+
+        # Recursively hook "record_args" to module and all its sub-modules.
+        # The function "record_args" records the inputs for each nn.Module,
+        # and later we will try exporting those nn.Module's with their recorded
+        # inputs.
+        def recursive_hook(module):
+            handle_pool.append(module.register_forward_pre_hook(record_args))
+            for name, sub in module._modules.items():
+                if isinstance(sub, torch.nn.ModuleList):
+                    for name1, sub1 in sub._modules.items():
+                        recursive_hook(sub1)
+                else:
+                    recursive_hook(sub)
+
+        exportable_list = {}
+
+        # Fill "exportable_list". exportable_list[module] = True means
+        # "module" can be wrapped as ORTModule. Otherwise, "module" is
+        # not exportable to ONNX.
+        def check_exportable(module):
+            sub_dict = module._modules
+            if not sub_dict:
+                # No sub-module exists, so this module is a leaf
+                # module in overall model hierarchy.
+                exportable = True
+                # Check if this leaf module is exportable.
+                for args in module_arg_pool[module]:
+                    try:
+                        torch.onnx.export(
+                            module, args, 'tmp.onnx', opset_version=9)
```

Comment:
hard-coding to tmp.onnx is not safe. Use https://docs.python.org/3/library/tempfile.html#tempfile.TemporaryFile instead

Reply:
No problem. Now we use
```python
                        try:
                            with tempfile.NamedTemporaryFile(prefix='sub-module') as temp:
                                torch.onnx.export(
                                    module, args, temp, opset_version=9)
                        except Exception as e:
```